internal/tlsmux/tls_mux.go

```diff
@@ -40,63 +46,67 @@ func (c *tlsMuxListener) Close() error {
 	return err
 }
 
-func New(logger logrus.FieldLogger, listener net.Listener, cert *x509.Certificate, tlsCert tls.Certificate) (net.Listener, net.Listener) {
-	var nonTlsConns = make(chan net.Conn, 128) // TODO decide on good buffer sizes for these channels
-	var nonTlsErrs = make(chan error, 128)
+func New(logger logrus.FieldLogger, listener net.Listener, cert *x509.Certificate, tlsCert tls.Certificate, keyLogWriter io.Writer) (net.Listener, net.Listener) {
+	var nonTLSConns = make(chan net.Conn, 128) // TODO decide on good buffer sizes for these channels
+	var nonTLSErrs = make(chan error, 128)
 	var tlsConns = make(chan net.Conn, 128)
 	var tlsErrs = make(chan error, 128)
 	go func() {
 		for {
 			rawConn, err := listener.Accept()
 			if err != nil {
-				nonTlsErrs <- err
+				nonTLSErrs <- err
 				tlsErrs <- err
 				continue
 			}
 
 			go func() {
 				conn := peekconn.New(rawConn)
 
-				isTls, err := conn.PeekMatch(tlsPattern, tlsPeekSize)
+				isTLS, err := conn.PeekMatch(tlsPattern, tlsPeekSize)
 				if err != nil {
-					nonTlsErrs <- err
+					nonTLSErrs <- err
 					tlsErrs <- err
 				}
-				if isTls {
-					handleTlsConn(logger, conn, cert, tlsConns)
+				if isTLS {
+					handleTLSConn(logger, conn, cert, tlsConns)
 				} else {
-					nonTlsConns <- conn
+					nonTLSConns <- conn
 				}
 			}()
 
 		}
 	}()
-
 	closer := &sync.Once{}
-	nonTlsListener := nonHTTPBouncer{
+	nonTLSListener := nonHTTPBouncer{
 		logger,
 		&tlsMuxListener{
 			Listener: listener,
 			close:    closer,
-			conns:    nonTlsConns,
+			conns:    nonTLSConns,
 		},
 		false,
 	}
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{tlsCert},
+		KeyLogWriter: keyLogWriter,
+	}
+	// Support HTTP/2: https://golang.org/pkg/net/http/?m=all#Serve
```

Comment:
As far as I understood, HTTP2 was already supported because `tlsMux` "unwraps" the TLS and the http server uses h2s to allow unencrypted HTTP2?
Was this not working for you?

Reply:
`net/http` package supports HTTP/2 out of the box if using TLS, but `tlsMux` uses `tls.NewListener` to wrap the raw listener, and then `proxy` uses `http.Serve` (and not `http.ServerTLS`). If I understand this correctly, this means that `net/http` is now unaware of the TLS part and therefore is not handling the HTTP/2 as expected.
I have tried to `http.ServerTLS` with "a plain" listener, and it seems to handle HTTP/2 well.

Anyway, it was not working for me, and this small change (following the docs) fixed it.